Handle unset optional properties (#47)
Materials-Consortia/OPTiMaDe: optimade.md

| 418  | Asked for a non-existent keyword                                      |
| 422  | Database returned (200) but the translator failed                     |

### <a name="h.3.3.4">3.3.4. Unset optional properties</a>

Unset optional properties in a database are properties that exist and have a specific value within a database for some materials entries,
but are undefined for other entries, i.e. have the value 'null' within a JSON file or SQL database.

> Unset properties are not returned in the response unless explicitly requested in the search query.

Reviewer: ```suggestion
Unset properties MUST NOT be returned in the response unless explicitly requested in the search query.
```

> Any comparisons involving unset properties should return a value of false,

Reviewer: ```suggestion
Any comparisons involving unset properties MUST be evaluated as false,
```

i.e. by definition the value of 'null' is outside of any defined search range.

If a property is explicitly requested in a search query without value range filters,

> then all entries otherwise satisfying the query should be returned, including those with null values for this property.

Reviewer: ```suggestion
then all entries otherwise satisfying the query SHOULD be returned, including those with null values for this property.
```

> The value of these properties should be set to the JSON 'null' value in the response.

Reviewer: ```suggestion
These properties MUST be set to the JSON 'null' value in the response.
```

Note that this only applies to properties that are described by a single value,
and not to properties with multiple attributes such as the structure.

> In this case, attributes that have null value are simply omitted from the returned structure.

Reviewer: In an OPTIMaDe meeting discussion, we note that we have some trouble understanding exactly what this means. What we think is needed is to say that the text in this section only describes how the api handles properties that themselves are null, and does not regulate which values inside the data structure used by a properties can be null. That is rather to be described in relation to the definition of those data structures.

So:
```suggestion
The text in this section describes how the api handles properties that are null. It does not regulate the handling of values inside property data structures that can be null. That use of null values is described in relation to the definition of those data structures.
```
(Note: I cannot get github to make a suggestion over several lines, but this is meant to replace the whole paragraph.)

to the corresponding database ID that was originally queried, using the object's
`base_url` value.

### <a name="h.3.3.4">3.3.4. Unset optional properties</a>

> Unset optional properties in a database are properties that exist and have a specific value within a database for some materials entries, but are undefined for other entries, i.e. have the value 'null' within a JSON file or SQL database.

Reviewer 2: To be consistent, I would always write null or 'null' as `null` instead, with backticks - both here and in a number of places below.

> Unset optional properties in a database are properties that exist and have a specific value within a database for some materials entries, but are undefined for other entries, i.e. have the value 'null' within a JSON file or SQL database.

Reviewer 2: I don't know if we want to mention SQL here. Maybe just say "i.e. have the value `null` or are missing in the data source (that can be a JSON file, a database, ...)" or something like this? (assuming I'm understanding what you want to say here)

Unset properties MUST NOT be returned in the response, unless explicitly requested in the search query. 

> Any comparisons involving unset properties MUST be evaluated as false,

Reviewer 2: Also here, I would use `false` instead of false?
